### crisp PCM の生成

- $n$: 基準/代替案数
- $V$: 一対比較値として取りうる値の集合 ($\{ 1/7, 1/5, 1/3, 1, 3, 5, 7 \}$ など)

1. $i < j$ について $a_{ij}$ を $V$ の中からランダムに選ぶ
2. $i > j$ について $a_{ij} = 1 / a_{ji}$ とする

### 区間 PCM の生成

- $n$: 基準/代替案数
- $N = \left\{ 1, 2, \dots, n \right\}$
- $r_{\max}$: 区間 PCM $A_k$ の各成分の幅を定めるときのパラメータ

crisp PCM $A = \left( a_{ij} \right)_{n \times n}$ を与える.

$\mathrm{DM}_k$ の区間 PCM $A_k = \left( A_{kij} \right)_{n \times n} = \left( \left[ a_{kij}^\mathrm{L}, a_{kij}^\mathrm{U} \right] \right)_{n \times n}$ を次のように与える.

1. $i < j$ の成分を次のように計算する.
    $$
    \begin{align*}
    a_{kij}^\mathrm{L} & = a_{ij} \exp \left( -r_{kij}^\mathrm{L} \right), \quad
    a_{kij}^\mathrm{U} & = a_{ij} \exp \left( r_{kij}^\mathrm{U} \right), \quad
    i, j \in N, ~~ i < j.
    \end{align*}
    $$
    ただし, $r_{kij}^\mathrm{L}, r_{kij}^\mathrm{U} ~ (k \in \{ 1,2 \}, ~~ i, j \in N, ~ i < j)$ は $0$ から $r_{\max}$ の一様分布に従う.

2. $i > j$ の成分は $A_{kij} = 1 / A_{kji} = \left[ \frac{1}{a_{kij}^\mathrm{U}}, \frac{1}{a_{kij}^\mathrm{L}} \right]$ とする.

### シミュレーション

1. 整合している crisp PCM $A_1, A_2$ を 1000 組生成する
2. crisp PCM 1 個に対して, 区間 PCM を 10 個生成する
3. 各区間 PCM について
    1. 手法 1 ~ 4 によって二重区間 PCM $\mathscr{A}^1, \mathscr{A}^2, \mathscr{A}^3, \mathscr{A}^4$ を計算
    2. 類似度を計算
        $$
        \mathrm{Sim}\left( A_{kij}, \mathscr{A}_{ij}^{l\pm} \right), ~~ i, j \in N, ~ i \neq j, ~ k \in \{ 1, 2 \}, ~ l \in \{ 1, 2, 3, 4 \}
        $$
    3. 類似度の算術平均をとる
        $$
        \rho^{l\pm} = \frac{1}{2n(n-1)} \sum_{k \in \{ 1,2 \}} \sum_{i, j \in N, i \neq j} \mathrm{Sim}\left( A_{kij}, \mathscr{A}_{ij}^{l\pm} \right), ~~ l \in \{ 1, 2, 3, 4 \}
        $$
3. $\rho^{l\pm}, ~ l \in \{ 1, 2, 3, 4 \}$ がそれぞれ 10000 個得られるので, その算術平均をとる

### 結果

PCM のサイズ別に表にしている.  
行が区間の幅に関する変数 $r_{\max}$ で, 下の行ほど幅が大きくなりうる.  
左 4 列が二重区間の内側で, 右 4 列が外側.


$n = 5$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0012|0.0000|0.0000|0.0000|0.0719|0.1531|0.1366|0.1677|
|0.5|0.0096|0.0000|0.0008|0.0002|0.2758|0.3452|0.2726|0.3718|
|1.0|0.0652|0.0032|0.0624|0.0194|0.4165|0.4156|0.3893|0.4727|
|1.6|0.1952|0.0322|0.2811|0.1651|0.4986|0.4340|0.4767|0.5181|

$n = 6$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0461|0.0902|0.0889|0.1139|
|0.5|0.0007|0.0000|0.0008|0.0002|0.1929|0.2596|0.1931|0.2833|
|1.0|0.0089|0.0000|0.0005|0.0001|0.3132|0.3359|0.2745|0.3859|
|1.6|0.0546|0.0004|0.0418|0.0053|0.3937|0.3631|0.4767|0.4393|

$n = 7$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0344|0.0606|0.0613|0.0806|
|0.5|0.0000|0.0000|0.0000|0.0000|0.1482|0.2021|0.1692|0.2282|
|1.0|0.0009|0.0000|0.0000|0.0000|0.2485|0.2808|0.2559|0.3289|
|1.6|0.0095|0.0000|0.0006|0.0000|0.3246|0.3125|0.3222|0.3876|

$n = 8$

|$r_{\max}$|$\rho^{1-}$|$\rho^{2-}$|$\rho^{3-}$|$\rho^{4-}$|$\rho^{1+}$|$\rho^{2+}$|$\rho^{3+}$|$\rho^{4+}$|
|---|---|---|---|---|---|---|---|---|
|0.1|0.0000|0.0000|0.0000|0.0000|0.0286|0.0483|0.0425|0.0616|
|0.5|0.0000|0.0000|0.0000|0.0000|0.1240|0.1710|0.1389|0.1981|
|1.0|0.0001|0.0000|0.0000|0.0000|0.2117|0.2454|0.2246|0.2963|
|1.6|0.0013|0.0000|0.0000|0.0000|0.2827|0.2775|0.2892|0.3532|


### 考察

- $n$ と一致度の関係
    - いずれのケースでも $n$ が大きくなると一致度が下がる
        - 制約が増え, 整合化が行われることがより多くなるため
- $r_{\max}$ と一致度の関係
    - いずれのケースでも $r_{\max}$ が大きくなると一致度が上がる
        - 区間が広くなると整合化が減る
- 内側の区間の一致度 $\rho^{l-}$ の手法間の比較
    - $n \geq 6$ では一致度 $0$ がほとんど
        - 区間の幅を広げたシミュレーションなどが必要 
    - 手法 1 がほとんどの場合最も良い
    - $n = 5, r_{\max} = 1.6$ では手法 3 が最も良く, 差があって手法 1, 4 と続く
        - これまでの数値例はこのケースに該当する
    - 区間の幅を広げていく ($r_{\max}$ を大きくしていく) と個人の重要度を先に求める手法 3, 4 の一致度が大きく改善する
- 外側の区間の一致度 $\rho^{l+}$ の手法間の比較
    - ほとんどのケースで一致度が手法  4 が最も良い
        - $n = 6, r_{\max} = 1.6$ のケースのみ手法 6 のほうが良い

---

In [1]:
println(Threads.nthreads())

22


In [2]:
using IntervalArithmetic
using Statistics

In [3]:
include("./crispPCM/index.jl")
include("./evaluation/index.jl")
include("./method1/index.jl")
include("./method2/index.jl")
include("./method3/index.jl")
include("./method4/index.jl")
include("./intervalPCM/index.jl")
include("./twofoldInterval/index.jl")
include("./twofoldIntervalPCM/index.jl")
include("./utils.jl")

R (generic function with 1 method)

In [4]:
Random.seed!(1)

TaskLocalRNG()

In [5]:
SimulationCase = @NamedTuple{
    CR::T,
    A::Matrix{T},
    A₁::Matrix{Interval{T}},
    A₂::Matrix{Interval{T}}
} where {T <: Real}

function generateSimulationCases(
    n::Integer, 
    values::Vector{T},
    width::T,
    numOfCrispPCM::Integer, 
    intervalPCMsPerCrispPCM::Integer
)::Array{SimulationCase, 1} where {T <: Real}

    simulation_cases = SimulationCase[]

    for _ in 1:numOfCrispPCM
        # Generate a Crisp PCM
        crisp_pcm = generateConsistentCrispPCM(n, values)
        CR = consistencyRatio(crisp_pcm)

        for _ in 1:intervalPCMsPerCrispPCM
            # Generate an Interval PCM for each Crisp PCM
            interval_pcm_1 = randamizedIntervalPCM(crisp_pcm, width)
            interval_pcm_2 = randamizedIntervalPCM(crisp_pcm, width)
            # Append the Crisp PCM and its corresponding Interval PCM as a tuple to the list
            push!(simulation_cases, (CR=CR, A=crisp_pcm, A₁=interval_pcm_1, A₂=interval_pcm_2))
        end
    end

    return simulation_cases
end


generateSimulationCases (generic function with 1 method)

In [6]:
SimulationResult = @NamedTuple{
    # A₁, A₂ の元となった crisp PCM
    A::Matrix{T},
    # 各 DM の与える区間 PCM
    A₁::Matrix{Interval{T}}, A₂::Matrix{Interval{T}},
    # method 1 ~ 4 の二重区間 PCM
    𝓐¹::Matrix{TwofoldInterval{T}}, 𝓐²::Matrix{TwofoldInterval{T}},
    𝓐³::Matrix{TwofoldInterval{T}}, 𝓐⁴::Matrix{TwofoldInterval{T}},
    # 𝓐ᵏ の二重区間の内側の区間
    𝓐¹⁻::Matrix{Interval{T}}, 𝓐²⁻::Matrix{Interval{T}},
    𝓐³⁻::Matrix{Interval{T}}, 𝓐⁴⁻::Matrix{Interval{T}},
    # 𝓐ᵏ の二重区間の外側の区間
    𝓐¹⁺::Matrix{Interval{T}}, 𝓐²⁺::Matrix{Interval{T}},
    𝓐³⁺::Matrix{Interval{T}}, 𝓐⁴⁺::Matrix{Interval{T}},
    # エラーメッセージ
    error::String
} where {T <: Real}

methodList = [method1, method2, method3, method4]

function runSimulation(
    case::SimulationCase{T}
    )::SimulationResult{T} where {T <: Real}
    𝓐 = Dict(); 𝓐⁻ = Dict(); 𝓐⁺ = Dict()
    
    try
        for k in 1:4
            𝓐[k] = methodList[k](case.A₁, case.A₂)
            𝓐⁻[k] = map(𝓐ᵢⱼ -> 𝓐ᵢⱼ[1], 𝓐[k])
            𝓐⁺[k] = map(𝓐ᵢⱼ -> 𝓐ᵢⱼ[2], 𝓐[k])
        end

        return (
            A=case.A, A₁=case.A₁, A₂=case.A₂,
            𝓐¹=𝓐[1], 𝓐²=𝓐[2], 𝓐³=𝓐[3], 𝓐⁴=𝓐[4],
            𝓐¹⁻=𝓐⁻[1], 𝓐²⁻=𝓐⁻[2], 𝓐³⁻=𝓐⁻[3], 𝓐⁴⁻=𝓐⁻[4],
            𝓐¹⁺=𝓐⁺[1], 𝓐²⁺=𝓐⁺[2], 𝓐³⁺=𝓐⁺[3], 𝓐⁴⁺=𝓐⁺[4],
            error=""
        )
    catch e
        n = size(case.A₁, 2)
        IPCM = fill(1..1, n, n)
        twofoldIPCM = fill((1..1, 1..1), n, n)
        return (
            A=case.A, A₁=case.A₁, A₂=case.A₂,
            𝓐¹=twofoldIPCM, 𝓐²=twofoldIPCM, 𝓐³=twofoldIPCM, 𝓐⁴=twofoldIPCM,
            𝓐¹⁻=IPCM, 𝓐²⁻=IPCM, 𝓐³⁻=IPCM, 𝓐⁴⁻=IPCM,
            𝓐¹⁺=IPCM, 𝓐²⁺=IPCM, 𝓐³⁺=IPCM, 𝓐⁴⁺=IPCM,
            error=string(e)
        )
    end
end

runSimulation (generic function with 1 method)

In [7]:
ConincidenceList = @NamedTuple{
    A₁_𝓐¹⁻::T, A₁_𝓐¹⁺::T, A₂_𝓐¹⁻::T, A₂_𝓐¹⁺::T,
    A₁_𝓐²⁻::T, A₁_𝓐²⁺::T, A₂_𝓐²⁻::T, A₂_𝓐²⁺::T,
    A₁_𝓐³⁻::T, A₁_𝓐³⁺::T, A₂_𝓐³⁻::T, A₂_𝓐³⁺::T,
    A₁_𝓐⁴⁻::T, A₁_𝓐⁴⁺::T, A₂_𝓐⁴⁻::T, A₂_𝓐⁴⁺::T,
    error::String
} where {T <: Real}

function calculateCoincidenceList(result::SimulationResult{T})::ConincidenceList{T} where {T <: Real}
    A₁ = result.A₁; A₂ = result.A₂
    𝓐¹⁻ = result.𝓐¹⁻; 𝓐¹⁺ = result.𝓐¹⁺
    𝓐²⁻ = result.𝓐²⁻; 𝓐²⁺ = result.𝓐²⁺
    𝓐³⁻ = result.𝓐³⁻; 𝓐³⁺ = result.𝓐³⁺
    𝓐⁴⁻ = result.𝓐⁴⁻; 𝓐⁴⁺ = result.𝓐⁴⁺
    try
        return (
            A₁_𝓐¹⁻=coincidenceIndex(A₁, 𝓐¹⁻), A₁_𝓐¹⁺=coincidenceIndex(A₁, 𝓐¹⁺),
            A₂_𝓐¹⁻=coincidenceIndex(A₂, 𝓐¹⁻), A₂_𝓐¹⁺=coincidenceIndex(A₂, 𝓐¹⁺),
            A₁_𝓐²⁻=coincidenceIndex(A₁, 𝓐²⁻), A₁_𝓐²⁺=coincidenceIndex(A₁, 𝓐²⁺),
            A₂_𝓐²⁻=coincidenceIndex(A₂, 𝓐²⁻), A₂_𝓐²⁺=coincidenceIndex(A₂, 𝓐²⁺),
            A₁_𝓐³⁻=coincidenceIndex(A₁, 𝓐³⁻), A₁_𝓐³⁺=coincidenceIndex(A₁, 𝓐³⁺),
            A₂_𝓐³⁻=coincidenceIndex(A₂, 𝓐³⁻), A₂_𝓐³⁺=coincidenceIndex(A₂, 𝓐³⁺),
            A₁_𝓐⁴⁻=coincidenceIndex(A₁, 𝓐⁴⁻), A₁_𝓐⁴⁺=coincidenceIndex(A₁, 𝓐⁴⁺),
            A₂_𝓐⁴⁻=coincidenceIndex(A₂, 𝓐⁴⁻), A₂_𝓐⁴⁺=coincidenceIndex(A₂, 𝓐⁴⁺),
            error=""
        )
    catch e
        return (
            A₁_𝓐¹⁻=NaN, A₁_𝓐¹⁺=NaN, A₂_𝓐¹⁻=NaN, A₂_𝓐¹⁺=NaN,
            A₁_𝓐²⁻=NaN, A₁_𝓐²⁺=NaN, A₂_𝓐²⁻=NaN, A₂_𝓐²⁺=NaN,
            A₁_𝓐³⁻=NaN, A₁_𝓐³⁺=NaN, A₂_𝓐³⁻=NaN, A₂_𝓐³⁺=NaN,
            A₁_𝓐⁴⁻=NaN, A₁_𝓐⁴⁺=NaN, A₂_𝓐⁴⁻=NaN, A₂_𝓐⁴⁺=NaN,
            error=string(e)
        )
    end
end

calculateCoincidenceList (generic function with 1 method)

In [8]:
function averageConincidenceLists(conincidenceLists::Vector{ConincidenceList{T}}) where {T <: Real}
    if length(conincidenceLists) == 0
        throw(ArgumentError("Empty list of ConincidenceLists"))
    end

    total_𝓐¹⁻ = zero(T); total_𝓐¹⁺ = zero(T)
    total_𝓐²⁻ = zero(T); total_𝓐²⁺ = zero(T)
    total_𝓐³⁻ = zero(T); total_𝓐³⁺ = zero(T)
    total_𝓐⁴⁻ = zero(T); total_𝓐⁴⁺ = zero(T)

    count = 0

    for list in conincidenceLists
        if list.error != "" continue end

        count += 1
        total_𝓐¹⁻ += list.A₁_𝓐¹⁻ + list.A₂_𝓐¹⁻
        total_𝓐¹⁺ += list.A₁_𝓐¹⁺ + list.A₂_𝓐¹⁺
        total_𝓐²⁻ += list.A₁_𝓐²⁻ + list.A₂_𝓐²⁻
        total_𝓐²⁺ += list.A₁_𝓐²⁺ + list.A₂_𝓐²⁺
        total_𝓐³⁻ += list.A₁_𝓐³⁻ + list.A₂_𝓐³⁻
        total_𝓐³⁺ += list.A₁_𝓐³⁺ + list.A₂_𝓐³⁺
        total_𝓐⁴⁻ += list.A₁_𝓐⁴⁻ + list.A₂_𝓐⁴⁻
        total_𝓐⁴⁺ += list.A₁_𝓐⁴⁺ + list.A₂_𝓐⁴⁺
    end
    count *= 2

    return (
        𝓐¹⁻ = total_𝓐¹⁻ / count, 𝓐¹⁺ = total_𝓐¹⁺ / count,
        𝓐²⁻ = total_𝓐²⁻ / count, 𝓐²⁺ = total_𝓐²⁺ / count,
        𝓐³⁻ = total_𝓐³⁻ / count, 𝓐³⁺ = total_𝓐³⁺ / count,
        𝓐⁴⁻ = total_𝓐⁴⁻ / count, 𝓐⁴⁺ = total_𝓐⁴⁺ / count,
    )
end

averageConincidenceLists (generic function with 1 method)

In [9]:
function runSimulate(
    n::Integer, 
    values::Vector{T},
    width::T,
    numOfCrispPCM::Integer, 
    intervalPCMsPerCrispPCM::Integer
    ) where {T <: Real}
    cases = generateSimulationCases(n, values, width, numOfCrispPCM, intervalPCMsPerCrispPCM)
    results = Vector{SimulationResult}(undef, length(cases))
    Threads.@threads for i in 1:length(cases)
        results[i] = runSimulation(cases[i])
    end
    conincidenceList = calculateCoincidenceList.(results)
    return averageConincidenceLists(conincidenceList)
end

runSimulate (generic function with 1 method)

In [10]:
values = [0.1, 0.5, 1.0, 1.6]

for n in 5:8
    println("n = $n")
    for value in values
        println(runSimulate(n, [1/9, 1/7, 1/5, 1/3, 1, 3, 5, 7, 9], value, 1000, 10))
    end
end

n = 5

(𝓐¹⁻ = 0.0012292655982896199, 𝓐¹⁺ = 0.07187403867569402, 𝓐²⁻ = 4.537924495082146e-16, 𝓐²⁺ = 0.1531243873253255, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.1366362230294785, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.16767902615045777)


(𝓐¹⁻ = 0.00958564623180895, 𝓐¹⁺ = 0.2758065125006352, 𝓐²⁻ = 3.7890027645501765e-5, 𝓐²⁺ = 0.34520356631972077, 𝓐³⁻ = 0.000764317217978492, 𝓐³⁺ = 0.2726270452104397, 𝓐⁴⁻ = 0.00018455351597496965, 𝓐⁴⁺ = 0.3718245965864604)


(𝓐¹⁻ = 0.06522011727308352, 𝓐¹⁺ = 0.4164834931865041, 𝓐²⁻ = 0.0032423626564446997, 𝓐²⁺ = 0.41562610584606047, 𝓐³⁻ = 0.06237200073595892, 𝓐³⁺ = 0.3892658658210833, 𝓐⁴⁻ = 0.019416684526763935, 𝓐⁴⁺ = 0.4726783093213293)


(𝓐¹⁻ = 0.19519911725524183, 𝓐¹⁺ = 0.4985628910940124, 𝓐²⁻ = 0.032161033885780614, 𝓐²⁺ = 0.43400317821650114, 𝓐³⁻ = 0.28108445061532666, 𝓐³⁺ = 0.47670995478819705, 𝓐⁴⁻ = 0.16505294873609058, 𝓐⁴⁺ = 0.5180635766985849)
n = 6

(𝓐¹⁻ = 4.297648716842587e-5, 𝓐¹⁺ = 0.046126685329609514, 𝓐²⁻ = 2.996105757958167e-16, 𝓐²⁺ = 0.09022726467427147, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.08891544530332435, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.11387364832792198)


(𝓐¹⁻ = 0.0007304865858008984, 𝓐¹⁺ = 0.192879995115428, 𝓐²⁻ = 9.984050880687179e-17, 𝓐²⁺ = 0.2595572185671469, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.19310131605065742, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.28325901751602456)


┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141
┌ Warning: Invalid input, empty interval is returned
└ @ IntervalArithmetic /opt/julia/packages/IntervalArithmetic/EquAX/src/intervals/intervals.jl:141


(𝓐¹⁻ = 0.008924689138638656, 𝓐¹⁺ = 0.31318322049136726, 𝓐²⁻ = 8.156300685753815e-17, 𝓐²⁺ = 0.3358839765807297, 𝓐³⁻ = 0.0005282482411123644, 𝓐³⁺ = 0.27449677420767277, 𝓐⁴⁻ = 7.538653173355916e-5, 𝓐⁴⁺ = 0.3859175199060911)


(𝓐¹⁻ = 0.05455882398744686, 𝓐¹⁺ = 0.3937285380601262, 𝓐²⁻ = 0.00039816027686887906, 𝓐²⁺ = 0.36314326331831237, 𝓐³⁻ = 0.04176260016576256, 𝓐³⁺ = 0.3615421073216701, 𝓐⁴⁻ = 0.005291573769343571, 𝓐⁴⁺ = 0.439253710831755)
n = 7

(𝓐¹⁻ = 7.67284716971498e-6, 𝓐¹⁺ = 0.03437594165630714, 𝓐²⁻ = 2.0204875936295446e-16, 𝓐²⁺ = 0.06056076192074657, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.06134265042343587, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.08055402297363452)


(𝓐¹⁻ = 4.032109383193416e-5, 𝓐¹⁺ = 0.14822204988070037, 𝓐²⁻ = 6.901567585207757e-17, 𝓐²⁺ = 0.2021183878742622, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.16920978391147404, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.22818904449013638)


(𝓐¹⁻ = 0.0009348718582917231, 𝓐¹⁺ = 0.2485196562632332, 𝓐²⁻ = 6.150134029685057e-17, 𝓐²⁺ = 0.2808287498506751, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2558601044149862, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.32891035790266226)


(𝓐¹⁻ = 0.00946220265836522, 𝓐¹⁺ = 0.32455376623547216, 𝓐²⁻ = 5.846893841942267e-17, 𝓐²⁺ = 0.31252492527004827, 𝓐³⁻ = 0.0006383137975704816, 𝓐³⁺ = 0.3222200160945192, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3876220219032487)
n = 8

(𝓐¹⁻ = 0.0, 𝓐¹⁺ = 0.028552343602762483, 𝓐²⁻ = 1.4275634583253184e-16, 𝓐²⁺ = 0.04827407644250388, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.042499953654163015, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.061603985384544396)


(𝓐¹⁻ = 1.243141551366449e-6, 𝓐¹⁺ = 0.1240488496952325, 𝓐²⁻ = 5.379823020375774e-17, 𝓐²⁺ = 0.17101934835738108, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.13893657246468463, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.19813452933092282)


(𝓐¹⁻ = 9.115277429253791e-5, 𝓐¹⁺ = 0.21169525044125143, 𝓐²⁻ = 4.8946302596376677e-17, 𝓐²⁺ = 0.24536746171032786, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2246172707967133, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.2962906961429553)


(𝓐¹⁻ = 0.0012802099046232485, 𝓐¹⁺ = 0.2827116687347921, 𝓐²⁻ = 4.8698573966670674e-17, 𝓐²⁺ = 0.2775177278262383, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2891726523037664, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3531951275421345)


### $n = 5$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 0.0012292655982896199, 𝓐¹⁺ = 0.07187403867569402, 𝓐²⁻ = 4.537924495082146e-16, 𝓐²⁺ = 0.1531243873253255, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.1366362230294785, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.16767902615045777)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 0.00958564623180895, 𝓐¹⁺ = 0.2758065125006352, 𝓐²⁻ = 3.7890027645501765e-5, 𝓐²⁺ = 0.34520356631972077, 𝓐³⁻ = 0.000764317217978492, 𝓐³⁺ = 0.2726270452104397, 𝓐⁴⁻ = 0.00018455351597496965, 𝓐⁴⁺ = 0.3718245965864604)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.06522011727308352, 𝓐¹⁺ = 0.4164834931865041, 𝓐²⁻ = 0.0032423626564446997, 𝓐²⁺ = 0.41562610584606047, 𝓐³⁻ = 0.06237200073595892, 𝓐³⁺ = 0.3892658658210833, 𝓐⁴⁻ = 0.019416684526763935, 𝓐⁴⁺ = 0.4726783093213293)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.19519911725524183, 𝓐¹⁺ = 0.4985628910940124, 𝓐²⁻ = 0.032161033885780614, 𝓐²⁺ = 0.43400317821650114, 𝓐³⁻ = 0.28108445061532666, 𝓐³⁺ = 0.47670995478819705, 𝓐⁴⁻ = 0.16505294873609058, 𝓐⁴⁺ = 0.5180635766985849)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 4.297648716842587e-5, 𝓐¹⁺ = 0.046126685329609514, 𝓐²⁻ = 2.996105757958167e-16, 𝓐²⁺ = 0.09022726467427147, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.08891544530332435, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.11387364832792198)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 0.0007304865858008984, 𝓐¹⁺ = 0.192879995115428, 𝓐²⁻ = 9.984050880687179e-17, 𝓐²⁺ = 0.2595572185671469, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.19310131605065742, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.28325901751602456)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.008924689138638656, 𝓐¹⁺ = 0.31318322049136726, 𝓐²⁻ = 8.156300685753815e-17, 𝓐²⁺ = 0.3358839765807297, 𝓐³⁻ = 0.0005282482411123644, 𝓐³⁺ = 0.27449677420767277, 𝓐⁴⁻ = 7.538653173355916e-5, 𝓐⁴⁺ = 0.3859175199060911)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.05455882398744686, 𝓐¹⁺ = 0.3937285380601262, 𝓐²⁻ = 0.00039816027686887906, 𝓐²⁺ = 0.36314326331831237, 𝓐³⁻ = 0.04176260016576256, 𝓐³⁺ = 0.3615421073216701, 𝓐⁴⁻ = 0.005291573769343571, 𝓐⁴⁺ = 0.439253710831755)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 7.67284716971498e-6, 𝓐¹⁺ = 0.03437594165630714, 𝓐²⁻ = 2.0204875936295446e-16, 𝓐²⁺ = 0.06056076192074657, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.06134265042343587, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.08055402297363452)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 4.032109383193416e-5, 𝓐¹⁺ = 0.14822204988070037, 𝓐²⁻ = 6.901567585207757e-17, 𝓐²⁺ = 0.2021183878742622, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.16920978391147404, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.22818904449013638)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 0.0009348718582917231, 𝓐¹⁺ = 0.2485196562632332, 𝓐²⁻ = 6.150134029685057e-17, 𝓐²⁺ = 0.2808287498506751, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2558601044149862, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.32891035790266226)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.00946220265836522, 𝓐¹⁺ = 0.32455376623547216, 𝓐²⁻ = 5.846893841942267e-17, 𝓐²⁺ = 0.31252492527004827, 𝓐³⁻ = 0.0006383137975704816, 𝓐³⁺ = 0.3222200160945192, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3876220219032487)

### $n = 6$

- $r_{\max} = 0.1$

    (𝓐¹⁻ = 0.0, 𝓐¹⁺ = 0.028552343602762483, 𝓐²⁻ = 1.4275634583253184e-16, 𝓐²⁺ = 0.04827407644250388, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.042499953654163015, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.061603985384544396)

- $r_{\max} = 0.5$

    (𝓐¹⁻ = 1.243141551366449e-6, 𝓐¹⁺ = 0.1240488496952325, 𝓐²⁻ = 5.379823020375774e-17, 𝓐²⁺ = 0.17101934835738108, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.13893657246468463, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.19813452933092282)

- $r_{\max} = 1.0$

    (𝓐¹⁻ = 9.115277429253791e-5, 𝓐¹⁺ = 0.21169525044125143, 𝓐²⁻ = 4.8946302596376677e-17, 𝓐²⁺ = 0.24536746171032786, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2246172707967133, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.2962906961429553)

- $r_{\max} = 1.6$

    (𝓐¹⁻ = 0.0012802099046232485, 𝓐¹⁺ = 0.2827116687347921, 𝓐²⁻ = 4.8698573966670674e-17, 𝓐²⁺ = 0.2775177278262383, 𝓐³⁻ = 0.0, 𝓐³⁺ = 0.2891726523037664, 𝓐⁴⁻ = 0.0, 𝓐⁴⁺ = 0.3531951275421345)